Sources/MongoSwift/MongoCursor.swift

```diff
@@ -56,34 +56,30 @@ public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
      *     invalid combination.
      */
     internal init(
+        stealing cursor: OpaquePointer,
+        connection: Connection,
         client: MongoClient,
         decoder: BSONDecoder,
         session: ClientSession?,
-        cursorType: CursorType? = nil,
-        initializer: (Connection) -> OpaquePointer
+        cursorType: CursorType? = nil
     ) throws {
-        let connection = try resolveConnection(client: client, session: session)
-        let cursor = initializer(connection)
         self.state = .open(cursor: cursor, connection: connection, client: client, session: session)
         self.cursorType = cursorType ?? .nonTailable
         self.decoder = decoder
         self.error = nil
 
-        if let err = self.getMongocError() {
-            // Errors in creation of the cursor are limited to invalid argument errors, but some errors are reported
-            // by libmongoc as invalid cursor errors. These would be parsed to .logicErrors, so we need to rethrow them
-            // as the correct case.
-            throw UserError.invalidArgumentError(message: err.errorDescription ?? "")
```

Comment:
I messaged about this on Slack before but I'm not sure we ever came to a resolution. but because of the fact that some mongoc methods returning cursors (like the list indexes one) do immediate I/O, this isn't actually true and we can see multiple types of errors show up here, for example, a server selection error. 

something we *could* do is in cases like `find` and `aggregate` where the method doesn't do I/O immediately, we could first see if there is an error and if so convert to invalidArgumentError, *then* trigger I/O, and if that errors then throw it as normal. 

but, in whatever the specific cases you are referring to in this comment are, that might lead to inconsistency in error reporting across cursor methods, because in the immediate I/O cases I think we would still throw those errors as logicErrors even if they should be converted as we can't tell where they came from?

Reply:
If you look at [`_mongoc_cursor_new_with_opts`](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/src/mongoc/mongoc-cursor.c#L230), pretty much every error there is an invalid argument error, but they're all returned as `MONGOC_ERROR_CURSOR_INVALID_CURSOR` errors. libmongoc also happens to use this domain/code for other things, such the error returned after the user iterates a dead cursor. For that reason, our parser in MongoError.swift returns a `logicError` when seeing that domain/code combo, and for the cursor creation specific cases, the original line just took the message out and put it in an argument error. 

I think what we can do is update `parseMongocError` to parse that combo as an argument error instead of a logic error, due to the fact that we now close cursors in Swift land and don't rely on mongoc to report errors for the iteration of dead cursors.

Reply:
that makes sense to me. changed.